src/coins.cpp

```diff
@@ -353,20 +382,425 @@ bool CCoinsViewCache::BatchWrite(CCoinsMap &mapCoins,
+        auto& iaMap = targetScInfo.mImmatureAmounts;
+
+        if (!iaMap.count(maturityHeight) )
+        {
+            // should not happen
+            LogPrint("sc", "ERROR %s():%d - scId=%s could not find immature balance at height%d\n",
+                __func__, __LINE__, scId.ToString(), maturityHeight);
+            return false;
+        }
+
+        LogPrint("sc", "%s():%d - immature amount before: %s\n",
+            __func__, __LINE__, FormatMoney(iaMap[maturityHeight]));
+
+        if (iaMap[maturityHeight] < entry.nValue)
+        {
+            // should not happen either
+            LogPrint("sc", "ERROR %s():%d - scId=%s negative balance at height=%d\n",
+                __func__, __LINE__, scId.ToString(), maturityHeight);
+            return false;
+        }
+
+        iaMap[maturityHeight] -= entry.nValue;
+        cacheSidechains[scId] = CSidechainsCacheEntry(targetScInfo, CSidechainsCacheEntry::Flags::DIRTY);
+
+        LogPrint("sc", "%s():%d - immature amount after: %s\n",
+            __func__, __LINE__, FormatMoney(iaMap[maturityHeight]));
+
+        if (iaMap[maturityHeight] == 0)
+        {
+            iaMap.erase(maturityHeight);
+            cacheSidechains[scId] = CSidechainsCacheEntry(targetScInfo, CSidechainsCacheEntry::Flags::DIRTY);
+            LogPrint("sc", "%s():%d - removed entry height=%d from immature amounts in memory\n",
+                __func__, __LINE__, maturityHeight );
+        }
+    }
+
+    // remove sidechain if the case
+    for(const auto& entry: tx.vsc_ccout)
+    {
+        const uint256& scId = entry.scId;
+
+        LogPrint("sc", "%s():%d - removing scId=%s\n", __func__, __LINE__, scId.ToString());
+
+        ScInfo targetScInfo;
+        if (!GetScInfo(scId, targetScInfo))
+        {
+            // should not happen
+            LogPrint("sc", "ERROR: %s():%d - scId=%s not in scView\n", __func__, __LINE__, scId.ToString() );
+            return false;
+        }
+
+        if (targetScInfo.balance > 0)
+        {
+            // should not happen either
+            LogPrint("sc", "ERROR %s():%d - scId=%s balance not null: %s\n",
+                __func__, __LINE__, scId.ToString(), FormatMoney(targetScInfo.balance));
+            return false;
+        }
+
+        cacheSidechains[scId] = CSidechainsCacheEntry(targetScInfo, CSidechainsCacheEntry::Flags::ERASED);
+
+        LogPrint("sc", "%s():%d - scId=%s removed from scView\n", __func__, __LINE__, scId.ToString() );
+    }
+    return true;
+}
+
+bool CCoinsViewCache::ApplyMatureBalances(int blockHeight, CBlockUndo& blockundo)
+{
+    LogPrint("sc", "%s():%d - blockHeight=%d, msc_iaundo size=%d\n", __func__, __LINE__, blockHeight,  blockundo.msc_iaundo.size() );
+
+    std::set<uint256> allKnowScIds;
+    queryScIds(allKnowScIds);
+
+    for(auto it_set = allKnowScIds.begin(); it_set != allKnowScIds.end(); ++it_set)
+    {
+        const uint256& scId = *it_set;
+        ScInfo info;
+        assert(GetScInfo(scId, info));
```

Comment:
ApplyMatureBalances is currently called at every block connection in order to mature the balance of sidechains before flushing it to the main chainstateDB cache, pcoinsTip.
So all the sidechains are loaded into the pcoinsTip, regardless whether the current block has any tx referring to any sidechain.
I wonder whether this could make the cache system less effective as the cache could be filled earlier thus increasing the frequency of flush to db (given current cache sizes).
We discussed with @alsala  about possible improvements.
One idea which has come up is to avoid maturing all sidechains at once "on write", i.e. before a flush.
Instead we could mature "on read", i.e. mature a specific sidechain balance on the moment it is requested from the chainstateDb, just before it is passed up to the cache system.
In this way we won't fill the cache with sidechains needing to be matured and still we will make sure that any processing of sidechains will still be on a mature one.